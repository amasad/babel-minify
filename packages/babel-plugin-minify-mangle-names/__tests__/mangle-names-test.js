jest.autoMockOff();

const plugin = require("../src/index");
const testUtils = require("../../../utils/test");

const transform = testUtils.transformer(plugin);
const transformModule = testUtils.transformer(plugin, {
  sourceType: "module"
});
const transformWithBlockScoping = testUtils.transformerWithBlockScoping(plugin);
const transformWithES2015 = testUtils.transformerWithES2015(plugin);

describe("mangle-names", () => {
  it("should not mangle names in the global namespace", () => {
    const source = `
      var Foo = 1;
    `;
    expect(transform(source)).toMatchSnapshot();
  });

  it("should mangle names", () => {
    const source = `
      function foo() {
        var xxx = 1;
        if (xxx) {
          console.log(xxx);
        }
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should handle name collisions", () => {
    const source = `
      function foo() {
        var x = 2;
        var xxx = 1;
        if (xxx) {
          console.log(xxx + x);
        }
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should be fine with shadowing", () => {
    const source = `
      var a = 1;
      function foo() {
        var xxx = 1;
        if (xxx) {
          console.log(xxx);
        }
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should not shadow outer references", () => {
    const source = `
      function bar() {
        function foo(a, b, c) {
          lol(a,b,c);
        }

        function lol() {}
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should mangle args", () => {
    const source = `
      function foo(xxx) {
        if (xxx) {
          console.log(xxx);
        }
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should ignore labels", () => {
    const source = `
      function foo() {
        meh: for (;;) {
          continue meh;
        }
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should not have labels conflicting with bindings", () => {
    const source = `
      function foo() {
        meh: for (;;) {
          var meh;
          break meh;
        }
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  // https://phabricator.babeljs.io/T6957
  xit("labels should not shadow bindings", () => {
    const source = `
      function foo() {
        var meh;
        meh: for (;;) {
          break meh;
        }
        return meh;
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should be order independent", () => {
    const source = `
      function foo() {
        function bar(aaa, bbb, ccc) {
          baz(aaa, bbb, ccc);
        }
        function baz() {
          var baz = who();
          baz.bam();
        }
        bar();
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should be order independent 2", () => {
    const source = `
      function foo() {
        (function bar() {
          bar();
          return function() {
            var bar = wow();
            bar.woo();
          };
        })();
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should handle only think in function scopes", () => {
    const source = `
      function foo() {
        function xx(bar, baz) {
          if (1) {
            yy(bar, baz);
          }
        }
        function yy(){}
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should be fine with shadowing 2", () => {
    const source = `
      function foo() {
        function xx(bar, baz) {
          return function(boo, foo) {
            bar(boo, foo);
          };
        }
        function yy(){}
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should not be confused by scopes", () => {
    const source = `
      function foo() {
        function bar() {
          var baz;
          if (baz) {
            bam();
          }
        }
        function bam() {}
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should not be confused by scopes (closures)", () => {
    const source = `
      function foo() {
        function bar(baz) {
          return function() {
            bam();
          };
        }
        function bam() {}
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should handle recursion", () => {
    const source = `
      function bar() {
        function foo(a, b, c) {
          foo(a,b,c);
        }
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should handle global name conflict", () => {
    const source = `
      function e() {
        function foo() {
          b = bar();
        }
        function bar() {}
      }
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should handle global name", () => {
    const source = `
      function foo() {
        var bar = 1;
        var baz = 2;
      }
    `;

    expect(transform(source, { blacklist: {foo: true, bar: true }})).toMatchSnapshot();
  });

  it("should handle deeply nested paths with no bindings", () => {
    const source = `
      function xoo() {
        function foo(zz, xx, yy) {
          function bar(zip, zap, zop) {
            return function(bar) {
              zap();
              return function() {
                zip();
              }
            }
          }
        }
      }
    `;
    expect(transform(source)).toMatchSnapshot();
  });

  it("should handle try/catch", () => {
    const source = `
      function xoo() {
        var e;
        try {} catch (e) {

        }
      }
    `;
    expect(transform(source)).toMatchSnapshot();
  });

  it("should not mangle vars in scope with eval" , () => {
    const source = `
      function foo() {
        var inScopeOuter = 1;
        (function () {
          var inScopeInner = 2;
          eval("inScopeInner + inScopeOuter");
          (function () {
            var outOfScope = 1;
          })();
        })();
      }
    `;
    expect(transform(source)).toMatchSnapshot();
  });

  it("should mangle names with local eval bindings", () => {
    const source = `
      function eval() {}
      function foo() {
        var bar = 1;
        eval('...');
      }
    `;
    expect(transform(source)).toMatchSnapshot();
  });

  it("should mangle names with option eval = true", () => {
    const source = `
      function foo() {
        var inScopeOuter = 1;
        (function () {
          var inScopeInner = 2;
          eval("...");
          (function () {
            var outOfScope = 1;
          })();
        })();
      }
    `;
    expect(transform(source, { eval: true })).toMatchSnapshot();
  });

  it("should integrate with block scoping plugin", () => {
    const source = `
      function f(x) {
        for (let i = 0; i; i++) {
          let n;
          if (n) {
            return;
          }
          g(() => n);
        }
      }
    `;

    expect(transformWithBlockScoping(source)).toMatchSnapshot();
  });

  it("should integrate with block scoping plugin 2", () => {
    const source = `
      (function () {
        function bar() {
          if (smth) {
            let entries = blah();
            entries();
          }
          foo();
        }
        function foo() { }
        module.exports = { bar };
      })();
    `;

    expect(transformWithBlockScoping(source)).toMatchSnapshot();
  });

  it("should keep mangled named consistent across scopes when defined later on", () => {
    const source = `
      (function() {
        function foo() {
          {
            var baz = true;

            {
              bar();
            }
          }
        }

        function bar() {}
      }());
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  it("should correctly mangle in nested loops", () => {
    const source = `
      (function () {
        for (let x in foo) {
          for (let y in foo[x]) {
            alert(foo[x][y]);
          }
        }
      })();
    `;

    expect(transform(source)).toMatchSnapshot();
  });

  // #issue55, #issue57
  it("should correctly mangle function declarations in different order", () => {
    const source = `
      (function(){
        (function() {
          for (let x in y) y[x];
          f(() => { g() });
        })();
        function g() {}
      })();
    `;

    expect(transformWithES2015(source)).toMatchSnapshot();
  });

  it("should NOT mangle functions & classes when keep_fnames is true", () => {
    const source = `
      (function() {
        class Foo {}
        const Bar = class Bar extends Foo {}
        var foo = function foo() {
          foo();
        }
        function bar() {
          foo();
        }
        bar();
        var baz = foo;
        baz();
      })();
    `;
    expect(transform(source, {keepFnames: true})).toMatchSnapshot();
  });

  it("should mangle variable re-declaration / K violations", () => {
    const source = `
      !function () {
        var foo = 1;
        foo++;
        var foo = 2;
        foo++;
      }
    `;
    expect(transform(source)).toMatchSnapshot();
  });

  it("should handle K violations - 2", () => {
    const source = `
      !function () {
        var bar = 1;
        bar--;
        var bar = 10;
        foo(bar)
        function foo() {
          var foo = 10;
          foo++;
          var foo = 20;
          foo(foo);
        }
      }
    `;
    expect(transform(source)).toMatchSnapshot();
  });

  it("should work with redeclarations", () => {
    const source = `
      (function () {
        var x = y;
        x = z;
        x;
      })();
    `;
    expect(transform(source)).toMatchSnapshot();
  });

  it("should reuse removed vars", () => {
    const source = `
      function Foo() {
        var a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
        var A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
        var $, _;
      }
    `;
    expect(transform(source)).toMatchSnapshot();
  });

  it("should mangle both referenced and binding identifiers with K violations", () => {
    const source = `
      (function () {
        var foo = bar,
            foo = baz;
        foo;
      })();
    `;
    expect(transform(source)).toMatchSnapshot();
  });

  it("should handle export declarations", () => {
    const source = `
      const foo = 1;
      export { foo };
      export const bar = 2;
      export function baz(bar, foo) {
        bar();
        foo();
      };
      export default function (bar, baz) {
        bar();
        baz();
      }
    `;
    expect(transformModule(source)).toMatchSnapshot();
  });

  it("should find global scope properly", () => {
    const source = unpad(`
      class A {}
      class B extends A {}
      (function () {
        class C {
          constructor() {
            new A();
            new B();
            C;
          }
        }
      })();
    `);
    const expected = unpad(`
      class A {}
      class B extends A {}
      (function () {
        class a {
          constructor() {
            new A();
            new B();
            a;
          }
        }
      })();
    `);
    expect(transform(source)).toBe(expected);
  });

  it("should mangle classes properly", () => {
    const source = unpad(`
      class A {}
      class B {}
      new A();
      new B();
      function a() {
        class A {}
        class B {}
        new A();
        new B();
      }
    `);
    const expected = unpad(`
      class A {}
      class B {}
      new A();
      new B();
      function a() {
        class b {}
        class c {}
        new b();
        new c();
      }
    `);
    expect(transform(source)).toBe(expected);
  });

  // https://github.com/babel/babili/issues/138
  it("should handle class exports in modules - issue#138", () => {
    const source = unpad(`
      export class App extends Object {};
    `);
    const expected = source;
    expect(transform(source, {}, "module")).toBe(expected);
  });
});
