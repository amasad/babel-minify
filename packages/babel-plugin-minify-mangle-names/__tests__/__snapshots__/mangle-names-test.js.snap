exports[`mangle-names should NOT mangle functions & classes when keep_fnames is true 1`] = `
"
(function () {
  class Foo {}
  const a = class Bar extends Foo {};
  var b = function foo() {
    foo();
  };
  function bar() {
    b();
  }
  bar();
  var c = b;
  c();
})();"
`;

exports[`mangle-names should be fine with shadowing 1`] = `
"
var a = 1;
function foo() {
  var b = 1;
  if (b) {
    console.log(b);
  }
}"
`;

exports[`mangle-names should be fine with shadowing 2 1`] = `
"
function foo() {
  function a(c, d) {
    return function (e, f) {
      c(e, f);
    };
  }
  function b() {}
}"
`;

exports[`mangle-names should be order independent 1`] = `
"
function foo() {
  function a(c, d, e) {
    b(c, d, e);
  }
  function b() {
    var c = who();
    c.bam();
  }
  a();
}"
`;

exports[`mangle-names should be order independent 2 1`] = `
"
function foo() {
  (function a() {
    a();
    return function () {
      var b = wow();
      b.woo();
    };
  })();
}"
`;

exports[`mangle-names should correctly mangle function declarations in different order 1`] = `
"\"use strict\";

(function () {
  (function () {
    for (var b in y) {
      y[b];
    }f(function () {
      a();
    });
  })();
  function a() {}
})();"
`;

exports[`mangle-names should correctly mangle in nested loops 1`] = `
"
(function () {
  for (let a in foo) {
    for (let b in foo[a]) {
      alert(foo[a][b]);
    }
  }
})();"
`;

exports[`mangle-names should handle K violations - 2 1`] = `
"
!function () {
  var b = 1;
  b--;
  var b = 10;
  a(b);
  function a() {
    var c = 10;
    c++;
    var c = 20;
    c(c);
  }
};"
`;

exports[`mangle-names should handle deeply nested paths with no bindings 1`] = `
"
function xoo() {
  function a(b, c, d) {
    function e(f, g, h) {
      return function (i) {
        g();
        return function () {
          f();
        };
      };
    }
  }
}"
`;

exports[`mangle-names should handle export declarations 1`] = `
"
const foo = 1;
export { foo };
export const bar = 2;
export function baz(a, b) {
  a();
  b();
};
export default function (a, b) {
  a();
  b();
}"
`;

exports[`mangle-names should handle global name 1`] = `
"
function foo() {
  var bar = 1;
  var a = 2;
}"
`;

exports[`mangle-names should handle global name conflict 1`] = `
"
function e() {
  function a() {
    b = c();
  }
  function c() {}
}"
`;

exports[`mangle-names should handle name collisions 1`] = `
"
function foo() {
  var a = 2;
  var b = 1;
  if (b) {
    console.log(b + a);
  }
}"
`;

exports[`mangle-names should handle only think in function scopes 1`] = `
"
function foo() {
  function a(c, d) {
    if (1) {
      b(c, d);
    }
  }
  function b() {}
}"
`;

exports[`mangle-names should handle recursion 1`] = `
"
function bar() {
  function d(e, f, g) {
    d(e, f, g);
  }
}"
`;

exports[`mangle-names should handle try/catch 1`] = `
"
function xoo() {
  var a;
  try {} catch (b) {}
}"
`;

exports[`mangle-names should ignore labels 1`] = `
"
function foo() {
  meh: for (;;) {
    continue meh;
  }
}"
`;

exports[`mangle-names should integrate with block scoping plugin 1`] = `
"
function f(a) {
  var b = function (d) {
    var e = void 0;
    if (e) {
      return {
        v: void 0
      };
    }
    g(() => e);
  };

  for (var d = 0; d; d++) {
    var c = b(d);
    if (typeof c === \"object\") return c.v;
  }
}"
`;

exports[`mangle-names should integrate with block scoping plugin 2 1`] = `
"
(function () {
  function a() {
    if (smth) {
      var c = blah();
      c();
    }
    b();
  }
  function b() {}
  module.exports = { bar: a };
})();"
`;

exports[`mangle-names should keep mangled named consistent across scopes when defined later on 1`] = `
"
(function () {
  function a() {
    {
      var c = true;

      {
        b();
      }
    }
  }

  function b() {}
})();"
`;

exports[`mangle-names should mangle args 1`] = `
"
function foo(a) {
  if (a) {
    console.log(a);
  }
}"
`;

exports[`mangle-names should mangle both referenced and binding identifiers with K violations 1`] = `
"
(function () {
  var a = bar,
      a = baz;
  a;
})();"
`;

exports[`mangle-names should mangle names 1`] = `
"
function foo() {
  var a = 1;
  if (a) {
    console.log(a);
  }
}"
`;

exports[`mangle-names should mangle names with local eval bindings 1`] = `
"
function eval() {}
function foo() {
  var a = 1;
  eval(\'...\');
}"
`;

exports[`mangle-names should mangle names with option eval = true 1`] = `
"
function foo() {
  var a = 1;
  (function () {
    var b = 2;
    eval(\"...\");
    (function () {
      var c = 1;
    })();
  })();
}"
`;

exports[`mangle-names should mangle variable re-declaration / K violations 1`] = `
"
!function () {
  var a = 1;
  a++;
  var a = 2;
  a++;
};"
`;

exports[`mangle-names should not be confused by scopes (closures) 1`] = `
"
function foo() {
  function a(c) {
    return function () {
      b();
    };
  }
  function b() {}
}"
`;

exports[`mangle-names should not be confused by scopes 1`] = `
"
function foo() {
  function a() {
    var c;
    if (c) {
      b();
    }
  }
  function b() {}
}"
`;

exports[`mangle-names should not have labels conflicting with bindings 1`] = `
"
function foo() {
  meh: for (;;) {
    var a;
    break meh;
  }
}"
`;

exports[`mangle-names should not mangle names in the global namespace 1`] = `
"
var Foo = 1;"
`;

exports[`mangle-names should not mangle vars in scope with eval 1`] = `
"
function foo() {
  var inScopeOuter = 1;
  (function () {
    var inScopeInner = 2;
    eval(\"inScopeInner + inScopeOuter\");
    (function () {
      var a = 1;
    })();
  })();
}"
`;

exports[`mangle-names should not shadow outer references 1`] = `
"
function bar() {
  function d(f, g, h) {
    e(f, g, h);
  }

  function e() {}
}"
`;

exports[`mangle-names should reuse removed vars 1`] = `
"
function Foo() {
  var aa, ba, ca, da, ea, fa, ga, ha, ia, ja, ka, la, ma, na, oa, pa, qa, ra, sa, ta, ua, va, wa, xa, ya, za;
  var Aa, Ba, Ca, Da, Ea, Fa, Ga, Ha, Ia, Ja, Ka, La, Ma, Na, Oa, Pa, Qa, Ra, Sa, Ta, Ua, Va, Wa, Xa, Ya, Za;
  var $a, _a;
}"
`;

exports[`mangle-names should work with redeclarations 1`] = `
"
(function () {
  var a = y;
  a = z;
  a;
})();"
`;
